### PR TITLE
Feat/erlang json parser

### PR DESCRIPTION
**Pull request by eltoncampos1**
@josevalim i saw your comments on the other pr and  Im reopend this pr to add JSON parser based on erlang `:json` module availability.

 - check :json availabilty otherwise jason is the parser
 - jason now is optional dependencies

### CODE REVIEW

**Review by josevalim — APPROVED** [review]

**Comment by zachallaun** [comment]
Unless I'm mistaken, change would be breaking for anyone on OTP prior to 27 who does not have an explicit `:jason` dependency.

In order not to break those cases, I think the `:jason` dependency would need to be specified like this:

```elixir
defp deps do
  [
    ...
  ] ++ maybe_jason()
end

defp maybe_jason do
  if System.otp_release() >= "27" do
    []
  else
    [{:jason, ">= 0.0.0"}]
  end
end
```

This would always depend on Jason for versions of OTP less than 27, and not depend on it at all for versions greater or equal.

**Comment by zachallaun** [comment]
> Hex does not really handle optional dependencies like that, it will either have it or not have it depending on the versionnused for publishing. 

Ah - that makes sense. Agreed that a line in the README (and CHANGELOG?) would do the trick.

**Comment by hauxir** [comment]
Can you add tests for nil(encode) and null(decode)?

Just something that caught my eye as I was playing around with :json
It doesn't seem to decode/encode null/nil correctly unless I'm mistaken.

> :json.encode(nil)
"nil"
>:json.decode("null")
:null

**Comment by eltoncampos1** [comment]
> Can you add tests for nil(encode) and null(decode)?
> 
> Just something that caught my eye as I was playing around with :json It doesn't seem to decode/encode null/nil correctly unless I'm mistaken.
> 
> > :json.encode(nil)
> > "nil"
> > :json.decode("null")
> > :null
> 
> add something like:
> 
> ```
> test "should encode a map with nil in it" do
>   assert JSON.encode!(%{foo: nil}) == "{\"foo\":null}"
> end
> 
> 
> test "should decode a map with null in it" do
>   assert JSON.decode!("[{\"foo\":null}]") == [%{"foo" => nil}]
> end
> ```

you right, i have to add some fixes to work properly with nullable values, thanks!

File: lib/flame/parser/json.ex [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      data
+      |> normalize_nullable()
+      |> :json.encode()
+      |> IO.iodata_to_binary()
+    end
+
+    def decode!(data) do
+      data
+      |> :json.decode()
+      |> normalize_nullable()
```
Comment (josevalim) [inline lib/flame/parser/json.ex:15]: There is no need to traverse it, you can use the options:

```
1> json:decode(<<"{\"foo\": null}">>, ok, #{null => nil}).
{#{<<"foo">> => nil},ok,<<>>}
```

File: lib/flame/parser/json.ex [hunk 2]
```diff
@@ -0,0 +1,37 @@
+defmodule FLAME.Parser.JSON do
+  @moduledoc false
+
+  if Code.ensure_loaded?(:json) do
+    def encode!(data) do
+      data
+      |> normalize_nullable()
+      |> :json.encode()
```
Comment (josevalim) [inline lib/flame/parser/json.ex:8]: There is no need to traverse it, you can use the options:

```
1> json:encode(#{<<"foo">> => nil}, fun(nil, _Encoder) -> <<"null">>; (Term, Encoder) -> json:encode_value(Term, Encoder) end).
["{",[[34,<<"foo">>,34],58|<<"null">>],[],"}"]
```

**Comment by eltoncampos1** [comment]
While testing, I observed a behavior difference between the :json library and the Jason library. Specifically, the return value from :json.encode was not ordered in the same way as with JSON.encode. However, when I tested it in iex, I noticed that the same value returned differently in the shell.

 
 OTP 27 elixir 1.17.2


Screenshot 2024-07-21 at 13 53 18

OTP 24, elixir 1.15.4


Screenshot 2024-07-21 at 13 55 17

**Comment by josevalim** [comment]
Maps do not guarantee an order of iteration, so these changes are expected. :)

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥
